src/lib/tabwidget/tabwidget.cpp

```diff
@@ -787,8 +787,17 @@ QByteArray TabWidget::saveState()
         if (!webTab)
             continue;
 
-        WebTab::SavedTab tab(webTab);
-        tabList.append(tab);
+        // save state does not support saving restore tab
```

Comment:
OK. How about this fix: (maybe when `tabList.isEmpty()` we should append a new tab?)
```diff
diff --git a/src/lib/tabwidget/tabwidget.cpp b/src/lib/tabwidget/tabwidget.cpp
index 18e4858..7975f03 100644
--- a/src/lib/tabwidget/tabwidget.cpp
+++ b/src/lib/tabwidget/tabwidget.cpp
@@ -782,16 +782,28 @@ QByteArray TabWidget::saveState()
 {
     QVector<WebTab::SavedTab> tabList;
 
+    int currentTabIndex = currentIndex();
+
     for (int i = 0; i < count(); ++i) {
         WebTab* webTab = weTab(i);
         if (!webTab)
             continue;
 
+        if (webTab->url().toString() == QLatin1String("qupzilla:restore")) {
+            if (i < currentTabIndex)
+                --currentTabIndex;
+            continue;
+        }
+
         WebTab::SavedTab tab(webTab);
         tabList.append(tab);
     }
 
     QByteArray data;
+
+    if (tabList.isEmpty())
+        return data;
+
     QDataStream stream(&data, QIODevice::WriteOnly);
 
     stream << tabList.count();
@@ -800,7 +812,7 @@ QByteArray TabWidget::saveState()
         stream << tab;
     }
 
-    stream << currentIndex();
+    stream << currentTabIndex;
 
     return data;
 }
```